posts/2020-11-27-typing-your-actors.adoc

```diff
@@ -0,0 +1,389 @@
+
+Initially I chose to use an untyped actor system intentionally, so we can actually talk about something that might be more useful, migrating your old Akka applications to a typed actor system. Why would you want to do that, well for me it is the type safety of not having everything being a function of `Any => Unit` and strong support to avoid anti-patterns (more eloquently explained https://www.youtube.com/watch?v=vF_-28I6yGE[here]), but feel free to search more if this migration is something for your project.
+
+== Overview
+
+Observable behaviour should stay the same after migration, which consists of:
+
+* A stream for each websocket connection with its own wire data https://github.com/lunatech-labs/lunatech-pointingpoker/blob/master/src/main/scala/com/lunatech/pointingpoker/websocket/WSMessage.scala#L10[domain] forwarding all incoming data to the main application controller (Room Manager) as well as providing an interface to forward outgoing data.
+* Room manager is responsible for translating from wire domain to the application https://github.com/lunatech-labs/lunatech-pointingpoker/blob/master/src/main/scala/com/lunatech/pointingpoker/Room.scala#L74-L84[domain] and https://github.com/lunatech-labs/lunatech-pointingpoker/blob/master/src/main/scala/com/lunatech/pointingpoker/RoomManager.scala#L56-L69[route] translated messages to appropriate rooms and manage each room lifecycle.
+* Rooms are representations of each planning poker session, holding all necessary https://github.com/lunatech-labs/lunatech-pointingpoker/blob/master/src/main/scala/com/lunatech/pointingpoker/Room.scala#L13-L15[information] and making sure that each participant is up to date with it.
+
+== Step by step
+
+First things first, we need to replace our old classical actor system with its typed counterpart:
+[source,scala]
+----
+implicit val system: ActorSystem[SpawnProtocol.Command] =
+    ActorSystem(Behaviors.setup[SpawnProtocol.Command](_ => SpawnProtocol()), "pointing-poker")
+----
+
+Takeaways:
+
+* There is no default guardian actor, you need to provide one.
+
+Tips:
+
+* Use https://doc.akka.io/api/akka/current/akka/actor/typed/SpawnProtocol$.html[SpawnProtocol], it allows to easily create new actors.
+
+
+Now that we have our actor system we need to create our top level actors:
+[source,scala]
+----
+system ! SpawnProtocol.Spawn(RoomManager(), "room-manager", Props.empty, system.ignoreRef)
+----
+
+Takeaways:
+
+* There is no `actorOf` method on a typed ActorSystem.
+* If you don't care/need the response, you can use `system.ignoreRef`
+
+
+I could have created the RoomManager during the `Behaviors.setup` when creating the actor system, but I also need the RoomManager reference in my http api, so I created it from “outside” using the `SpawnProtocol.Spawn` which will reply with the ActorRef once created. However as you can see, I’m ignoring the ref for now, because to actually use it, we need to talk how to _ask_ information out of the actor system:
```

Comment:
```suggestion
I could have created the RoomManager during the `Behaviors.setup` when creating the actor system, but I also need the RoomManager reference in my http api, so I created it from the “outside” using the `SpawnProtocol.Spawn` which will reply with the ActorRef once created. However as you can see, I’m ignoring the ref for now, because to actually use it, we need to talk how to _ask_ information out of the actor system:
```

Reply:
Also the last sentence is it missing the word first? or am I misunderstanding it?
```suggestion
I could have created the RoomManager during the `Behaviors.setup` when creating the actor system, but I also need the RoomManager reference in my http api, so I created it from “outside” using the `SpawnProtocol.Spawn` which will reply with the ActorRef once created. However as you can see, I’m ignoring the ref for now, because to actually use it, we need to first talk about how to _ask_ information from the actor system:
```

Reply:
**_http api_** should be all uppercase: **_HTTP API_**